src/Kestrel.Transport.Abstractions/Internal/TransportConnection.cs

```diff
@@ -55,6 +61,44 @@ public TransportConnection()
 
         public CancellationToken ConnectionClosed { get; set; }
 
+        public CancellationToken ConnectionClosing { get; set; }
+
+        public void TickHeartbeat(DateTimeOffset now)
+        {
+            lock (_heartbeatLock)
+            {
+                if (_heartbeatHandlers == null)
+                {
+                    return;
+                }
+
+                foreach (var (handler, state) in _heartbeatHandlers)
+                {
+                    handler(now, state);
+                }
+            }
+        }
+
+        public void OnHeartbeat(Action<object> action, object state)
+        {
+            // REVIEW: We could avoid this allocation with 2 lists
```

Comment:
This would be solved pretty easily by the caching ISystemClock suggestion.

Reply:
Yep, I like this!